### ISSUE TITLE
SSL_new returns NULL because openssl fails to re-open /dev/Xrandom when in chroot.

### ISSUE THREAD

**Issue by EmericBr** [openssl/openssl#5330]
This bug was not observed on previous versions and seems introduced with commits:

08a65d9686b131cb4193feaaf1d5cef941fa349c and a93ba405b0327db9106f8f224112a2b64fb264e7

**Comment by mattcaswell** [comment]
Ping @mspncp. Thoughts?

**Comment by mspncp** [comment]
@EmericBr, could you please provide a little bit more information like the operating system and the context in which you encounter the problem?  When you write `/dev/Xrandom` do you mean "`/dev/random` or `/dev/urandom`", or are you talking about something else?

Essentially the two commits you mentioned changed the reseeding policy, they reduced both the amount of RNG output and elapsed time after which a reseeding occurs. So your main problem is that `/dev/Xrandom` is not not mounted in the chroot environment, the commits only make the problem occur earlier: Since the random device is missing, the reseed fails due to lack of entropy input.

Is it feasible for you to mount the random device in the chroot environment? Another workaround would be to recompile OpenSSL, changing the random source by adding the

```
--with-rand-seed=[getrandom|devrandom|os|egd|rdcpu|librandom]
```
option to the configure command. IIRC, this option can be specified multiple times. I just noticed that `perl Configure --help`does not mention the option (this should be fixed), but you will find the option if you look into the Configure script, and it is mentioned in `man RAND_add`.

Alternatively, you could have a look at pull request #5314 by @kroeckx which adds support for the `getrandom()` system call on linux if available. If you have a linux system, you could give it a try. Since that random source is not bound to a file, you don't have the problems when chrooting.

Hope that helps!

**Comment by mspncp** [comment]
>  I just noticed that perl Configure --helpdoes not mention the option (this should be fixed)

Correction, the Configure script does mention `[--with-xxx[=vvv]]`. @levitte, I assume this brevity is on purpose? What is the recommended way then to find out about all options that are supported?

**Comment by EmericBr** [comment]
Hi @mspcnp,

I wrote  /dev/Xrandom because an strace shows that many entries are opened in /dev (and all of them fail):

connect(8, {sa_family=AF_INET, sin_port=htons(443), sin_addr=inet_addr("192.168.122.101")}, 16) = -1 EINPROGRESS (Operation now in progress)
open("/dev/urandom", O_RDONLY)          = -1 ENOENT (No such file or directory)
open("/dev/random", O_RDONLY)           = -1 ENOENT (No such file or directory)
open("/dev/srandom", O_RDONLY)          = -1 ENOENT (No such file or directory)
open("/dev/urandom", O_RDONLY)          = -1 ENOENT (No such file or directory)
open("/dev/random", O_RDONLY)           = -1 ENOENT (No such file or directory)
open("/dev/srandom", O_RDONLY)          = -1 ENOENT (No such file or directory)
setsockopt(8, SOL_SOCKET, SO_LINGER, {onoff=1, linger=0}, 8) = 0
close(8)                                = 0

The #5314 : "Use both getrandom() and /dev/urandom by default"

Does it mean that by default "/dev/urandom" is used by default and setting --with-rand-seed=getrandom gives the warranty that any random files  will be opened ?

Regards,
Emeric

**Comment by mspncp** [comment]
My understanding is the following:


You can add one or more `--with-rand-seed=xxx` options to explicitely enable random sources. The option `--with-rand-seed=xxx` will effectively add the `#define OPENSSL_RAND_SEED_XXX`. If you don't specify any such option, it defaults to `--with-rand-seed=os`. 

Now on master (without PR #5314) `OPENSSL_RAND_SEED_OS` activates only `OPENSSL_RAND_SEED_DEVRANDOM` in rand_unix.c (provided that `DEVRANDOM` is defined). This means, you have to add

```
--with-rand-seed=devrandom  --with-rand-seed=getrandom
```
to have both random sources.


PR #5314 adds an automatic detection, such that for a default configuration (`OPENSSL_RAND_SEED_OS`) both options `OPENSSL_RAND_SEED_DEVRANDOM` and `OPENSSL_RAND_SEED_GETRANDOM` are selected automatically, if available. So this means you don't have to add any extra option

BTW: The constant DEVRANDOM is a comma separated list which is #defined in e_os.h. This is the list of random sources which are tried, as you can see in the strace.

```
#  if defined(__s390__)
#   define DEVRANDOM "/dev/prandom","/dev/urandom","/dev/hwrng","/dev/random"
#  else
#   define DEVRANDOM "/dev/urandom","/dev/random","/dev/srandom"
#  endif
# endif
```

**Comment by richsalz** [comment]
One correction: its a comma-separated list, not a repeated flag.

**Comment by mspncp** [comment]
> One correction: its a comma-separated list, not a repeated flag.

Ok, thanks! Good to know :smiley:

**Comment by mspncp** [comment]
I updated https://github.com/openssl/openssl/issues/5330#issuecomment-364965630.

**Comment by kroeckx** [comment]
After my patch, on systems where getrandom() is availble, it
should only use getrandom() unless that fails for some reason.
It's trying all sources until it has enough entropy.

I would also like to see support for syscalls on other OSs that
provide it, and use them by default. So PR for that are always
welcome.

**Comment by mspncp** [comment]
@EmericBr, were you able to solve your problem? If yes, I'd also be interested to hear which of the proposed variants you chose. By the way, PR #5314 has been merged to master meanwhile, so you just could pull a fresh copy, if that variant worked for you.

**Comment by mattcaswell** [comment]
Can this be closed now?

**Comment by mspncp** [comment]
@EmericBr, thanks for reporting this issue. I assume everything turned out fine, since there was no answer for a while. So I'm closing this ticket now. If you still have problems, you are free to reopen it or to create a new issue.

**Comment by EmericBr** [comment]
I solve my problem patching openssl to force the use of syscall getrandom.

I needed to patch because getrandom is available only since glibc 2.25. and i should declare the SYS_getrandom manually.

I think a lot of distro with glibc < 2.25 will be affected by this is issue.

Is there any way to keep /dev/urandom open to avoid a re-open during runtime (after the chroot?)

**Comment by kroeckx** [comment]
The latest beta release should use the system call if glibc doesn't support it. Did you try per6?

**Comment by lukastribus** [comment]
As far as I understand there are 2 distinct issues here:
- the commits Emeric mentioned break existing applications in chroot, if getrandom() is not available
- getrandom() may or may not be automatically enabled depending on build issues

Yes, getrandom() is amazing and fixes a lot of problems including this one, but unless support for non-getrandom() capable systems is dropped completely in OpenSSL 1.1.1, we will still have to make sure the fallback code is not broken, even if the application is in chroot.

Am I wrong?

**Comment by kroeckx** [comment]
There is also support for the getrandom syscall without using the glibc wrapper. It should work on kernels >= 3.17 (released 2014-10-05). You need support for it at build time by having header files that define the syscall number, and then actually running on a kernel that supports the syscall. I understand that @EmericBr's problem is that his include files are too old but that his kernel supports it. I think in that case he really should install the header files from the correct kernel version. I currently don't see how we can improve this.

If your kernel doesn't support getrandom(), I suggest you create /dev/urandom in the chroot.

**Comment by lukastribus** [comment]
So what you are saying is that you are willing to break all applications/setups in OpenSSL 1.1.1 out there that do not have getrandom(), do not have /dev/xrandom in chroot and that worked fine for decades by initializing OpenSSL before chroot()?

Is that what you are saying?

Just so we are clear: this means hanging/crashing/malfunctioning applications with OpenSSL 1.1.1, not immediately, but after some time.

[Haproxy 1.8 with OpenSSL 1.1.1-pre4 stops working after 1 hour](https://www.mail-archive.com/haproxy@formilux.org/msg29592.html) is what a report looks like today because of this behavior.

Are you suggesting unix daemons using SSL and chroot should take care of creating /dev/xrandom in the chroot or are you suggesting they ignore this problem?

**Comment by richsalz** [comment]
Right now we are opting for strong security.  This means older kernels without the syscall, should link /dev/random into the chroot of any servers they run.  We should document this more clearly.
(It used to be that you would have to do this, albeit rarely, anyway. But the CSPRNG is much stronger with this release, and it requires re-seeding more often.)

I will re-open this issue and mark it for 1.1.1 so that someone will put a note into the README or INSTALL file.  Or is a FAQ entry good enough?

**Comment by lukastribus** [comment]
I think you are underestimating the breakage this will cause - documented or not. At the very least abort() and crash the application, so that coming to this conclusion is not such a nightmare.

And the question remains - since openssl is a library, not an application - what do you expect applications do to about this? Ignore it? Mess with the chroot itself? Should the application call abort() itself? What would your suggestion for a webserver like nginx, or a proxy like haproxy be?

**Comment by richsalz** [comment]
I think perhaps you are overestimating it. :)  But neither of us will ever really know.  The idea of doing an abort if the file isn't available is an interesting one.  Perhaps @kroeckx wants to do that?   Perhaps @mspncp can add some docs?  Or, upgrade the OS to support getrandom and configure openssl to use it?

And by the way, thanks for struggling through helping us improve a pre-release version!  I mean that seriously.

**Comment by mspncp** [comment]
> What would your suggestion for a webserver like nginx, or a proxy like haproxy be?

The application (e.g the webserver) is expected to check the return value on every call of RAND_bytes() and not take randomness for granted. If the call fails (which is be a recoverable error if the failure of the entropy source is temporary, but is unrecoverable in the given case were getrandom() is not available and no /dev/urandom device was provided in the changeroot environment), the application is expected to cancel the operation for which it was requesting the random bytes (e.g, the key exchange) and issue a proper error message. Not more, not less.

If a server application like nginx intends to chroot, it could add /dev/urandom before dropping privileges and chrooting, or alternatively instruct the user (in the documention) how to setup the changeroot environment properly.

> The idea of doing an abort if the file isn't available is an interesting one. 

Only the application itself has the right to abort(), and it should always be the last resort. _Never ever_ should a library abort() without consent of the application, except maybe for a severe internal programming error. Aborting the application from a library is vandalism, it deprives the application of the ability to handle the error appropriately. It even may destroy user data. Just imagine a GUI application which simply bombs out and disappears without letting the user save his work.

**Comment by kroeckx** [comment]
The reason I started adding the direct syscall version and not rely on the glibc version, and also adding support for the FreeBSD's and OpenBSD's system calls is that I actually feared this to be a big problem. And I hope providing the syscalls will at least fix this for a very large amount of people, but clearly not all people.

The RNG needs to be reseeded. Maybe the timers aren't ideal. Setting them longer will just make it last longer before it breaks. Setting it shorter will make it use more entropy, which probably isn't ideal, but it will show the error sooner.

I guess you would actually like to know this as soon as possible after start, with a clear message what is wrong. We could go and do something like check that /dev/urandom is available 10 seconds after the RNG is initialized. I don't like abort, but maybe syslog()?

**Comment by mspncp** [comment]
The application could call RAND_poll() followed by RAND_status() after chrooting if it wants to know immediately, whether reseeding still works.

**Comment by mspncp** [comment]
Enabling reseeding is a breaking change, admittedly. But the same holds for disabling an unsecure cipher. In both cases security has precedence over compatibility. A random generator which never reseeds is not considered secure anymore these days.

The only compromise I see is that we could add a configure option or a config file entry to set the reseed interval (where setting it to 0 means disabling it).

**Comment by mattcaswell** [comment]
> The only compromise I see is that we could add a configure option or a config file entry to set the reseed interval (where setting it to 0 means disabling it).

This sounds plausible. We have things like "enable-weak-ssl-ciphers" now so it's not without precedent to add this sort of thing. As long as its accompanied with sufficiently sternly worded advice not to do it unless you understand the consequences. :-)

**Comment by t8m** [comment]
Thanks to the raw syscall support I really wonder what are any remaining non-exotic environments which are affected. But yes, I can imagine that if you are doing local build of openssl-1.1.1 for let's say Red Hat Enterprise Linux 6 that you could make use of such option. The other option could be to keep the /dev/urandom file descriptor open from the first call - libgcrypt does this.

**Comment by mspncp** [comment]
> This sounds plausible. We have things like "enable-weak-ssl-ciphers" now so it's not without precedent to add this sort of thing. 

What I don't like about the configure option is that it affects all applications, even those that don't chroot and don't need this feature. So a per-application solution would be better. Even then, disabling the `master_reseed_interval` and `master_reseed_time_interval` unconditionally via `RAND_DRBG_set_reseed_defaults()` might be unnecessary and weaken security unnecessarily. A better solution might be to have a function called `RAND_DRBG_enable_careless_reseeding()` (or `RAND_DRBG_ignore_reseeding_failures()`, or `[your suggestion here]`) which attempts to reseed but returns without error (cancelling the reseeding without entering the error state, instead leaving the drbg instantiated and resetting only the `master_reseed[_time]_interval`) if the entropy source fails.

**Comment by mspncp** [comment]
> The other option could be to keep the /dev/urandom file descriptor open from the first call - libgcrypt does this.

Sounds good at first glance, but is it fork-safe to reuse the file descriptor for the random device in the child processes? (CC: @kaduk, @paulidale)

**Comment by mspncp** [comment]
Side note: it seems like BoringSSL is using a urandom file descriptor, protected by CRYPTO_LOCK_RAND, see [crypto/rand/urandom.c:80](https://boringssl.googlesource.com/boringssl/+/2272/crypto/rand/urandom.c#80).

**Comment by lukastribus** [comment]
The configure option is not really useful imho, as most of the times it is not accessible to the user. Disabling reseeding via API or switching to a "lazy" mode via API is way better.

Of course, keeping the FD open is the best of all the options, in my opinion. I like it very much because it keeps periodic reseeding fully working in all situation.


It is true that we don't have those issues in CentOS 7, Debian or Ubuntu stable. But calling everything exotic that is not a maintstream Linux distribution is a rather dangerous mindset to put yourself into if you are developing a portable library.

Kernel headers are not available when cross-compiling, and vendor's may be forced to use older kernels for HW support reasons in the embedded or industrial sector. It is not as black and white. By the way the 3.16 kernel has a projected kernel.org EOL of April, 2020.

**Comment by EmericBr** [comment]
Sorry, i'm late in the discussion.

In my case kernel is > 3.17 but glibc is < 2.25. I'm using a cross compilation and don't have the kernel header into the sysroot . libc headers are enought for all the other packaged software and i found curious to need kernel header just for openssl.

As Lukas point, to keep the FD open would be the best solution in my opinion. Honestly i was completely surprised to see that openssl retry to re-open that one during runtime.

**Comment by richsalz** [comment]
I really dislike keeping an FD open.  How about the application open it, and a new RAND_seed_fd() API is added?

**Comment by t8m** [comment]
Because the issue is relevant mostly for legacy systems I am afraid that changing the applications to call the API would be practically as difficult as changing the applications to have /dev/urandom in the chroot. So I do not think this is good solution. And it even gives a possibility for the application to improperly seed the RNG by opening something else than /dev/urandom.

**Comment by kroeckx** [comment]
I'm not sure keeping the fd open actually fixes it in all cases,
but it would at least be better.

**Comment by mspncp** [comment]
> Or are you saying that headers also don't define SYS_getrandom because they don't match the kernel you're running?

Kurt, would it make sense to attempt a raw syscall even if `SYS_getrandom` isn't defined  (by defining the constant ourselves)? The syscall numbers only get added and never change, right? So wouldn't the worst case just be that the syscall number is unassigned in an older kernel and `syscall()` would just fail with `errno=ENOSYS`?

**Comment by kroeckx** [comment]
As far as I know, syscall numbers are arch dependent.

**Comment by kroeckx** [comment]
On i386 it's 355, on x86_64 it's 318 ...

**Comment by davidben** [comment]
> it seems like BoringSSL is using a urandom file descriptor, protected by CRYPTO_LOCK_RAND

That's a pretty old version of the code (the old locking stuff's been gone for ages), but we do keep a urandom fd open when using urandom, yes. Though Chromium passes an fd down to BoringSSL before entering the sandbox. We do not make `/dev/urandom`, or anything other file, available to the renderers inside the sandbox. (Actually, how would you expose `/dev/urandom` without the rest of `/dev`? Exposing all of `/dev` to a sandbox seems a bad plan.) RDRAND or `getrandom` we call directly and ensure the sandbox allows the latter.

Though these sort of sandboxes tend to fixate on little implementation details, as shown in this bug. I'm not sure what's the right way for a library like OpenSSL to support them. On the Chromium side, we do not like to call into system libraries within a sandbox at all. Where we do, it's with careful evaluation. (E.g. libc is hard to avoid, but it is usually "obvious" about its syscall usage.)

System libraries update independently of your application. A sandbox is a very different environment from the usual system, so it's likely outside any compatibility guarantees of that library. For instance, if one were to write a syscall filter for OpenSSL 1.1.0, it may break in OpenSSL 1.1.1, when it starts using `getrandom`. Is that a compatibility break? I don't know, did OpenSSL promise that its syscall pattern would forever stay the same? If no, system OpenSSL can't support syscall sandboxes. If yes, 1.1.1 can't add `getrandom` which isn't good for other consumers.

**Comment by t8m** [comment]
The set of syscalls called is definitely not part of public API/ABI, this is internal detail. BTW glibc can change the called syscalls on new versions as well. And nobody tries to say that glibc is not backwards ABI compatible.

**Comment by lukastribus** [comment]
I agree that this is a corner case and that there are numerous workarounds to overcome this problem for the admin, once the issue is found.

I disagree that coming to that conclusion needs to take hours/days of painstaking analysis and googling around (or by harassing application developers on support channels), which is why I believe either the library or the application needs to fail early.

Since there is no consensus about how to fix the root cause in the library, and because I believe failing early in the library is also difficult (if not more difficult), the conclusion for me is that the application needs to detect this and fail early itself.

So based on @mspncp suggestion, after chrooting and when OpenSSL is >= 1.1.1, based on the RAND_poll() return code I'm failing the application:

```
#include <openssl/rand.h>
/*
 * chrooting here
 */
#if (OPENSSL_VERSION_NUMBER >= 0x10101000L) && !defined(LIBRESSL_VERSION_NUMBER) && !defined(OPENSSL_IS_BORINGSSL)
	if (! RAND_poll()) {
		ha_alert("Fatal error: OpenSSL unable to reseed entropy pool after chroot (required by OpenSSL 1.1.1 due to periodic reseeding).\n");
		exit(1);
	}
#endif
```


This works for me. Is this something you'd agree an application should do? Other thoughts?

**Comment by richsalz** [comment]
That makes sense.  Perhaps we could turn that into a FAQ?

**Comment by lukastribus** [comment]
Yes, I agree that the FAQ can list something like the snippet above. But given that FAQ is not included in the source distribution, important information's don't belong there imho (it's fine for snippets and workarounds like this one though).

I think we will need a small warning in CHANGES as well as a more concrete information in the RAND_DRBG_reseed man page (including the default reseed values, which are currently undocumented).

I will see if I can put together a few sentences in the coming days.

**Comment by mspncp** [comment]
> I think we will need a small warning in CHANGES as well as a more concrete information in the RAND_DRBG_reseed man page ...

Go ahead, that's a good idea. IMHO the optimal place for a detailed documentation might be the [RESEEDING](https://github.com/openssl/openssl/blob/master/doc/man7/RAND_DRBG.pod#reseeding) section of the RAND_DRBG(7) manpage. In other places were appropriate (e.g. RAND_DRBG_reseed, RAND_poll), you can then mention the chroot problem and refer to RAND_DRBG(7).

In particular, it would be a good idea to add a caveat to the following optimistic statement in RAND(7):

> As a normal application developer, you don't have to worry about any details, just use RAND_bytes(3) to obtain random data. Having said that, there is one important rule to obey: Always check the error return value of RAND_bytes(3) and don't take randomness for granted.

([RAND(7) - DESCRIPTION](https://github.com/openssl/openssl/blob/master/doc/man7/RAND.pod#description))


> ... (including the default reseed values, which are currently undocumented).

I remember that in the beginning when I created the RAND_DRBG documentation, I intended to document the default values officially, but then I decided that the precise values are an implementation detail. Only the fact that periodic reseeding occurs is relevant. Having to adjust the default values in two different locations (source and documentation) is error prone. So maybe just mention the names of the constants and [where they are defined](https://github.com/openssl/openssl/blob/master/crypto/rand/rand_lcl.h#L27)?

**Comment by lukastribus** [comment]
Before we continue going down the doc road I just want to emphasize that just because I'm suggesting doc improvements does not mean I agree with giving up on finding a technical solution within the library. I still do not think that documenting this and requiring the application to fail early is the best coarse of action, even worse is letting the end-user figure this out by himself (without early-failure).

I also don't feel like there has been a discussion about why keeping the FD open is a bad idea (as opposed to open/and close it once an hour), or why that would not cover this case.

Before 1.1.1 hit beta, at least 4 different people from different backgrounds hit this and reported it either here or to the application developers. I would be very surprised if this would not come up numerous times when 1.1.1 hits stable. Sure, in -pre6 we also use raw getrandom() syscalls, which will improve coverage significantly, but still, I doubt the number will be small. chroot() and a <=3.16 is not *THAT* exotic.


Having said that, short of a technical solution within the library I'm suggesting the following doc adjustments, for which I can submit a PR in the next days if there's rough consensus:


```
diff --git a/CHANGES b/CHANGES
index 7a47852..0db73dc 100644
--- a/CHANGES
+++ b/CHANGES
@@ -154,7 +154,9 @@
      a port of the default random generator from the OpenSSL FIPS 2.0
      object module. It is a hybrid deterministic random bit generator
      using an AES-CTR bit stream and which seeds and reseeds itself
-     automatically using trusted system entropy sources.
+     automatically using trusted system entropy sources. Inaccessible system
+     entropy sources can lead to permanent failures after some time, due to
+     automatic reseeding. Please see RAND_DRBG(7) manpage for further details.
 
      Some of its new features are:
       o Support for multiple DRBG instances with seed chaining.
diff --git a/doc/man3/RAND_DRBG_reseed.pod b/doc/man3/RAND_DRBG_reseed.pod
index da3a40b..e609361 100644
--- a/doc/man3/RAND_DRBG_reseed.pod
+++ b/doc/man3/RAND_DRBG_reseed.pod
@@ -61,6 +61,7 @@ and the reseed time interval
 of DRBG instances.
 The default values are set independently for master DRBG instances (which don't
 have a parent) and slave DRBG instances (which are chained to a parent DRBG).
+Default values defined in I<crypto/rand/rand_lcl.h> apply otherwise.
 
 =head1 RETURN VALUES
 
diff --git a/doc/man7/RAND.pod b/doc/man7/RAND.pod
index 578018f..9bacd93 100644
--- a/doc/man7/RAND.pod
+++ b/doc/man7/RAND.pod
@@ -24,10 +24,11 @@ to be initialized ('seeded') explicitly.
 It seeds and reseeds itself automatically using trusted random sources
 provided by the operating system.
 
-As a normal application developer, you don't have to worry about any details,
+As a normal application developer, you don't have to worry about most details,
 just use L<RAND_bytes(3)> to obtain random data.
-Having said that, there is one important rule to obey: Always check the error
-return value of L<RAND_bytes(3)> and don't take randomness for granted.
+Having said that, there are a few important rules to obey: Always check the error
+return value of L<RAND_bytes(3)>, don't take randomness for granted and make
+sure system entropy sources can be accessed at all times (especially after chroot).
 
 For long-term secrets, you can use L<RAND_priv_bytes(3)> instead.
 This method does not provide 'better' randomness, it uses the same type of CSPRNG.
diff --git a/doc/man7/RAND_DRBG.pod b/doc/man7/RAND_DRBG.pod
index 9f7f124..d2a4e40 100644
--- a/doc/man7/RAND_DRBG.pod
+++ b/doc/man7/RAND_DRBG.pod
@@ -156,7 +156,6 @@ A DRBG instance seeds itself automatically, pulling random input from
 its entropy source. The entropy source can be either a trusted operating
 system entropy source, or another DRBG with access to such a source.
 
-Automatic reseeding occurs after a predefined number of generate requests.
 The selection of the trusted entropy sources is configured at build
 time using the --with-rand-seed option. The following sections explain
 the reseeding process in more detail.
@@ -183,6 +182,15 @@ the DRBG is seeding or reseeding.
 The last case ensures that the DRBG automatically recovers
 from the error as soon as the entropy source is available again.
 
+If entropy sources are permanently unavailable, there is no way for DRBG
+to recover. Special care has to be taken by applications when I<chroot()>ing
+into an environment without I</dev/[urandom|srandom|random]> and when
+I<getrandom()> or equivalent syscalls are not available: as the initial
+seed may happen before I<chroot()>, the application will initially work, but
+when either I<reseed_interval> or I<reseed_time_interval> trigger, DRBG
+reseeding will permanently fail. An application may catch this scenario
+early by checking the return code off I<RAND_poll()> after I<chroot()>ing.
+
 =head2 Manual Reseeding
 
 In addition to automatic reseeding, the caller can request an immediate

```


Let me know if there's anything in there you don't like.

**Comment by mspncp** [comment]
Thank you for the patch, Lukas. I think it is consensus that _something_ should be documented, the question is only about the details. So would you mind submitting your additions as a pull request right away, regardless of the cla question? This would facilitate reviewing them and keeping the discussion about them in a separate thread.

**Comment by mspncp** [comment]
Note: I will be travelling today, so I won't be able to have a look at it before the weekend.

**Comment by mspncp** [comment]
> I also don't feel like there has been a discussion about why keeping the FD open is a bad idea (as opposed to open/and close it once an hour), or why that would not cover this case.

I'm not opposed in general, it's still an option. But it would only mitigate the maintenance problem if it would be enabled unconditionally. A new api call or a configur{e,ation} option would still leave us with breaking applications and people asking us what to do about it. So someone would have to come up with a pr that implements it properly for all corner cases (across forks, without sanitizers complaining about open file handles, etc.). And then we would need to convince Rich :-).

**Comment by kroeckx** [comment]
I don't think the problem is after a fork. If they fork inside the
chroot, they have the same problem with 1.1.0 and probably already
set up /dev/urandom inside the chroot. But as we've seen, it might
actually be with the wrong permissions, and that it still depends
on when it drops root, before or after they the RNG is initialized.

**Comment by mspncp** [comment]
I didn‘t mean opening /dev/urandom after fork. I just meant managing the lifetime of the handle properly across forks without leaving handle leaks at exit. Not that it‘s difficult, but it needs to be done properly, since we‘re a general purpose library and not an application.

**Comment by paulidale** [comment]
The only negative I see from keeping a FD around is the resource use (i.e. one FD plus kernel resources).  This will work across forks and will be thread safe.  If the read fails (which is extremely unlikely if not impossible), the FD would need to be closed and reopened which hits the same problems.

**Comment by mspncp** [comment]
@paulidale, maybe you want to give it a try? ;-)

**Comment by kroeckx** [comment]
If nobody else does this, I will try to write a patch to keep the fd open.

**Comment by mspncp** [comment]
> If nobody else does this, I will try to write a patch to keep the fd open.

The effort of writing the code for keeping the fd open is imho negligible compared to the effort to test it (e.g to implement a chroot() test).

**Comment by paulidale** [comment]
Is a test case using `chroot(2)` even possible using the existing infrastructure. The  `chroot(2)` system call requires privileges.  It might be possible to write a test where the file is removed and the lack of error noted.  Likewise for the failure case when the `read(2)` fails and the file cannot be reopened.

**Comment by mspncp** [comment]
A semi-manual solution would be to implement the test in such a way that it terminates without failing  if it does not have the necessary capabilities to chroot(). The test would simply issue a warning and provide instructions how to enable it. For example by calling a script `sudo tests/enable-chroot-test` which adds the necessary capabilites to the binary. It's not ideal, but better than nothing.

**Comment by paulidale** [comment]
Could the persistent FD be made optional?  A configure option `--work-in-chroot` or something more specific?

**Comment by mspncp** [comment]
If that option would be disabled by default, the main problem would not be solved, namely that chroot applications compiled for 1.1.0 would break when the system OpenSSL library of the linux distribution gets updated from 1.1.0 to 1.1.1. The application would not fail immediately but only after a while and the unprepared administrator would have a hard time figuring out what the problem is.

Also, a configure option would not help unless the administrator replaces the distribution's system library by his own configured and compiled version. This sounds like much more effort for the administrator than just adding a urandom device to the chroot environment.

**Comment by mspncp** [comment]
On the other side: it's hard for me to imagine a linux distribution that keeps an old kernel version around and at the same time updates to a brand new OpenSSL version.

**Comment by kroeckx** [comment]
I'm also not sure how to test this on a system that does have getentropy()

**Comment by mspncp** [comment]
You would need to configure it exclusively using `--with-rand-seed=devrandom`

**Comment by mspncp** [comment]
Should be fixed now, since #6432 has been merged (c7504aeb640a).

**Comment by lukastribus** [comment]
Thanks everyone!

**Comment by levitte** [comment]
> Are you suggesting unix daemons using SSL and chroot should take care of creating /dev/xrandom in the chroot ... ?

I'm currently looking at the aftermath of the changes that were made to solve the "chroot problem", and I would say that yeah, they should. Or the manager that set up the chrooted environment should. This is chroot environment setup 101, you do need to setup the stuff that the application running in the chrooted environment needs.
